### ISSUE TITLE
Feature Request - Programatically show/hide tooltip

### ISSUE THREAD

**Issue by sospirited** [recharts/recharts#1231]
## Do you want to request a feature or report a bug?
Feature.

## What is the current behavior?

The available callbacks for a chart (e.g. `LineChart`), are:

For tooltips to appear, the user hovers on the chart component with it's child `<Tooltip>` component and internal callbacks are fired to set the `<Tooltip>` prop to active and provide it with props for the `payload`, which it uses to style and display the hover tooltip.

To my knowledge, this is the only way to make tooltips appear, via user interaction with the chart.

## What is the expected behavior?

As this is a feature request, I will express the desired behaviour.

I would like a way to programatically display the tooltip on my chart.
E.g. triggered via a callback on another parent component of the `<Chart>` component), where I provide dataset key I want the tooltip to display information for.

For example, if we look at the SimpleLineChart - http://recharts.org/#/en-US/examples/SimpleLineChart

I would like to be able to call something like `showTooltip('Page D')`, on the `<LineChart>` to programatically make my tooltip appear.

### The 'why' factor
This way, if I have other components showing relevant contextual info relating to `Page D` on another component, when the user interacts/hovers on THAT component, I can fire this new callback to show the tooltip on the chart at the same time, providing a better user experience. 

It need not get in the way of lifecycle events as user's mouse interaction with the chart can still go through it's normal events and ultimately be hidden once the user's mouse leaves the chart again.

In summary, I would like to request API methods on the charts for:
`showTooltip()`
`hideTooltip()` 

## Which versions of Recharts, and which browser / OS are affected by this issue? Did this work in previous versions of Recharts?
latest

**Comment by knowbody** [comment]
is there a plan on making this happen?

**Comment by VictoriaGribkova** [comment]
I face with this issue too. I want to show the tooltip if item is focused. This is necessary for better accessibility. But there is no feature to programmatically show and hide the tooltip.

**Comment by luisgurmendezMLabs** [comment]
What happens if you pass the active prop yourself?

**Comment by luisgurmendez** [comment]
What happens if you control the active prop yourself?

Just tryed to do this, and it seems that the Tooltip overrides the active prop..
So what I thought was to create a CustomTooltip and pass it through the content prop of the Tooltip

`<Tooltip content={<CustomTooltip customActive={active} />}  />`

```
import DefaultTooltipContent from 'recharts/lib/component/DefaultTooltipContent';
.
.
.

interface CustomTooltipProps extends TooltipProps{
    customActive: boolean;
}

const CustomTooltip: React.FC<CustomTooltipProps> =({customActive, ...rest}) => {
  if(!customActive){
      return null
   }
  return <DefaultTooltipContent {...rest} />;
}
```

And if needed :

```
declare module 'recharts/lib/component/DefaultTooltipContent' {
    import React from 'react';
    import { TooltipProps } from 'recharts';

    declare const DefaultTooltipContent: React.ComponentType<TooltipProps>;

    export default DefaultTooltipContent;
}

```

**Comment by MatiasCS** [comment]
I need this too!😢

**Comment by nasmithan** [comment]
I was able to solve this by passing a component and some css. When the "recharts-tooltip-wrapper" div is empty (`label === 'IGNORE'`), I set display to none. Here's a simple version of what I'm doing:
Recharts tooltip:
```
<Tooltip
  content={<CustomTooltip />}
/>
```

Custom content component:
```
export function CustomTooltip({
  payload,
}) {
  if (payload.length === 0 || payload[0].payload.label === 'IGNORE') return null;
  return payload[0].payload.label;
}
```

CSS:
```
'.recharts-tooltip-wrapper:empty': {
  display: 'none',
}
```

**Comment by rashmimh** [comment]
> I was able to solve this by passing a component and some css. When the "recharts-tooltip-wrapper" div is empty (`label === 'IGNORE'`), I set display to none. Here's a simple version of what I'm doing:
> Recharts tooltip:
> 
> ```
> <Tooltip
>   content={<CustomTooltip />}
> />
> ```
> 
> Custom content component:
> 
> ```
> export function CustomTooltip({
>   payload,
> }) {
>   if (payload.length === 0 || payload[0].payload.label === 'IGNORE') return null;
>   return payload[0].payload.label;
> }
> ```
> 
> CSS:
> 
> ```
> '.recharts-tooltip-wrapper:empty': {
>   display: 'none',
> }
> ```

Many many thanks to you.

**Comment by calvinx408** [comment]
I was able to solve this by using onMouseEnter/onMouseLeave per each div chart and just passing state to the customtooltip

**Comment by DmitriyKochergin** [comment]
After digging sources I was able to programmatically hide tooltip by hacking the state of BarChart component via ref. It was needed to hide tooltips on mobile devices when device orientation changes.

barChartRef = React.createRef();
...
<BarChart ref={this.barChartRef}
...

this.barChartRef.current.setState({isTooltipActive: false});

...

**Comment by corysimmons** [comment]
We could do so much stuff if `index` and `activeIndex` (hovered or clicked) was just added to everything. `index`es are on a random handful of things in Recharts. :(

**Comment by DarthVitalus** [comment]
> After digging sources I was able to programmatically hide tooltip by hacking the state of BarChart component via ref. It was needed to hide tooltips on mobile devices when device orientation changes.
> 
> barChartRef = React.createRef(); ... <BarChart ref={this.barChartRef} ...
> 
> this.barChartRef.current.setState({isTooltipActive: false});
> 
> ...

How do you pass the data item to tooltip? To control payload and position?

**Comment by DarthVitalus** [comment]
thanks to @DmitriyKochergin !
with little trick, we can achieve expected behavior ;)
https://codesandbox.io/s/recharts-programmatically-show-tooltip-bhfnwt?file=/src/App.tsx

**Comment by AndreSilva1993** [comment]
> thanks to @DmitriyKochergin ! with little trick, we can achieve expected behavior ;) https://codesandbox.io/s/recharts-programmatically-show-tooltip-bhfnwt?file=/src/App.tsx

Is this a supported way though? Or is this solution dependent on internal implementation details that could change at all times?

**Comment by ckifer** [comment]
Eh definitely not "officially" supported. But we'll try our best not to break anything outside of a major version bump

**Comment by mgifford** [comment]
A tooltip to build on that's quite accessible is:
 https://designsystem.digital.gov/components/tooltip/
 
 or for that matter:
 https://getbootstrap.com/docs/4.1/components/tooltips/

**Comment by NereonNeo** [comment]
> thanks to @DmitriyKochergin ! with little trick, we can achieve expected behavior ;) https://codesandbox.io/s/recharts-programmatically-show-tooltip-bhfnwt?file=/src/App.tsx

Thank you for codesanbox example, thanks so mach

**Comment by kyle-marshall** [comment]
https://github.com/recharts/recharts/issues/784#issuecomment-2166896963

I'm not sure if this works for all chart types, but I found a simple solution which is working for me using BarChart.
(First I tried the solutions on this thread but only got partial success / was left with some undesired flickering / briefly incorrectly highlighted items)

**Comment by letkeeko** [comment]
In v2, chartRef?.current.setState({ isTooltipActive: false }) worked fine.

In v3, it throws: Uncaught TypeError: chartRef?.current.setState is not a function.

This was really helpful for our fully custom tooltip (which includes a close button, per our designer's request).

Haven't found a straightforward way to replicate this "closable tooltip" behaviour in v3...

**Comment by ckifer** [comment]
> chartRef?.current.setState 

This wasn't ever intended to be a publicly accessible API. Pretty much just an accident and then a hack that all of the categorical chart state was ever accessible via that ref due to how class components work. 

You can use `active={myIsActiveState}` to toggle the tooltip on and off

**Comment by ckifer** [comment]
https://stackblitz.com/edit/react-fqjf9jmq?file=src%2FExample.tsx&showSidebar=0

**Comment by letkeeko** [comment]
@ckifer Thanks, really appreciate the example - straightforward. In our current implementation, the tooltip is handled internally when it opens, which is default behaviour. The only thing missing is the ability to programmatically close it. If I refactor our code to your example, how can I access the internal tooltip state? Hoping to keep using the default Recharts tooltip open behaviour - just need a way to close it programmatically :)

**Comment by ckifer** [comment]
This is a way to close it programmatically. The tooltip can either be controlled or not controlled - in this case controlled by you in a way that is the same as default tooltip behavior (you don't need to use click if you want the normal hover functionality). Then just set state to false when you want to close it. 

We don't have any internal "closeTooltip" functions. The behavior of the tooltip is based upon what the tooltip trigger is, the coordinate where your mouse is, if active is true or false (internally or externally if set via props), and if there is a tooltip payload to show at that location
